File: packages/lambda/md-mcp/src/handler.ts
import { handler as saveMarkdown } from './save_markdown.js';
import { handler as loadMarkdown } from './load_markdown.js';
import { handler as editMarkdown } from './edit_markdown.js';
import type {
  SaveMarkdownInput,
  LoadMarkdownInput,
  EditMarkdownInput,
} from './models.js';

interface LambdaContext {
  clientContext?: {
    custom?: {
      bedrockAgentCoreToolName?: string;
    };
  };
}

export const handler = async (event: unknown, context: LambdaContext) => {
  const toolName = context.clientContext?.custom?.bedrockAgentCoreToolName ?? '';
  const action = toolName.includes('___') ? toolName.split('___').pop() : toolName;

  switch (action) {
    case 'save_markdown':
      return saveMarkdown(event as SaveMarkdownInput);
    case 'load_markdown':
      return loadMarkdown(event as LoadMarkdownInput);
    case 'edit_markdown':
      return editMarkdown(event as EditMarkdownInput);
    default:
      throw new Error(`Unknown tool: ${toolName}`);
  }
};

File: packages/lambda/pdf-mcp/src/handler.py
"""Main handler for PDF MCP Lambda - routes based on tool name from context."""

from .create_pdf import create_pdf
from .extract_tables import extract_tables
from .extract_text import extract_text


def handler(event: dict, context) -> dict:
    """Route to appropriate handler based on tool name from context.

    The tool name is provided via context.client_context.custom['bedrockAgentCoreToolName']
    in the format: "{target_name}___{tool_name}" (e.g., "pdf-mcp___extract_text")

    Args:
        event: Tool-specific input parameters
        context: Lambda context with client_context containing tool name

    Returns:
        Result from the appropriate handler
    """
    tool_name = context.client_context.custom.get("bedrockAgentCoreToolName", "")
    # Extract action from tool name (format: "pdf-mcp___extract_text")
    action = tool_name.split("___")[-1] if "___" in tool_name else tool_name

    if action == "extract_text":
        return extract_text(event)
    elif action == "extract_tables":
        return extract_tables(event)
    elif action == "create_pdf":
        return create_pdf(event)
    else:
        raise ValueError(f"Unknown tool: {tool_name}")
